### PR TITLE
Support USP sequence parallel attention for eagle3 training

### PR DESCRIPTION
**Pull request by uygnef**
This PR updates USP training to use SGLang-produced, pre-sharded Mooncake tensors instead of loading a full sample on one
  training rank and scattering it during prefetch.

  The main changes are:

  - Store USP rank-local shards in Mooncake from the SGLang producer side.
  - Fan out USP sample metadata to all ranks in the corresponding SP group.
  - Make each training rank read only its own `{mooncake_key}_usp{sp_rank}` tensors.
  - Reconstruct local `loss_mask`, `attention_mask`, and `position_ids` in the USP data fetcher.
  - Fix a USP zero-loss local shard case that could skip attention backward collectives.
  - Add coverage for the local-zero-loss/global-nonzero-loss USP shard case.

  ## Motivation

  The previous USP path loaded full Mooncake tensors in the training prefetch path and then distributed local shards across SP
  ranks. This made prefetch do extra distributed communication and exposed a collective ordering issue when one USP shard had no
  local loss tokens while another shard in the same Ulysses group did.

  With this change, SGLang writes per-SP-rank tensors directly to Mooncake. Training workers only receive lightweight metadata
  through Ray queues and independently load their own shard.

  ## Data Flow 
 ```mermaid 
 flowchart TD
      A[SGLang generate / prefill] --> B[hidden_states, input_ids, last_hidden_states]
      B --> C[EagleMooncakeStore.put_usp_shards]

      C --> D{for each sp_rank}
      D --> E[split_usp_batch]
      E --> F0[Mooncake key: key_usp0<br/>input_ids shard<br/>hidden_states shard<br/>target/lhs shard]
      E --> F1[Mooncake key: key_usp1<br/>input_ids shard<br/>hidden_states shard<br/>target/lhs shard]
      E --> FN[Mooncake key: key_uspN<br/>input_ids shard<br/>hidden_states shard<br/>target/lhs shard]

      C --> G[InferenceOutput<br/>mooncake_key=key<br/>tensor_shapes=global shapes<br/>packed_loss_mask<br/>metadata:
  usp_sharded=true]

      G --> H[AsyncInferenceManager<br/>merge metadata]
      H --> I[AsyncTrainingController]
      I --> J{DP rank's SP group}
      J --> Q0[Queue for train rank sp0]
      J --> Q1[Queue for train rank sp1]
      J --> QN[Queue for train rank spN]

      Q0 --> R0[Trainer rank sp0]
      Q1 --> R1[Trainer rank sp1]
      QN --> RN[Trainer rank spN]

      R0 --> S0[read Mooncake key_usp0]
      R1 --> S1[read Mooncake key_usp1]
      RN --> SN[read Mooncake key_uspN]

      S0 --> T0[reconstruct local loss_mask<br/>attention_mask<br/>position_ids]
      S1 --> T1[reconstruct local loss_mask<br/>attention_mask<br/>position_ids]
      SN --> TN[reconstruct local loss_mask<br/>attention_mask<br/>position_ids]

      T0 --> U[USP forward/backward]
      T1 --> U
      TN --> U
```
  ## Implementation Details

  ### SGLang / Mooncake

  - Added EagleMooncakeStore.put_usp_shards(...).
  - The producer splits input_ids, hidden_states, and last_hidden_states / target by USP rank.
  - Shards are stored under rank-local keys:
      - {mooncake_key}_usp0_*
      - {mooncake_key}_usp1_*
      - ...
  - SGLang sets metadata={"usp_sharded": True} on USP outputs.
  - SGLang patch files are updated so patched SGLang calls put_usp_shards(...) when TORCHSPEC_USP_SHARDED_MOONCAKE=1.

  ### Controller

  - USP training now creates one queue per SP rank.
  - The controller fans out each sample metadata object to all ranks in the DP rank's SP group.
  - Inference metadata is merged with original sample metadata so fields like has_thinking and usp_sharded are both preserved.

  ### Data Fetcher

  - USP mode now only supports sharded Mooncake samples.
  - Each rank reads:
      - f"{mooncake_key}_usp{sp_rank}"
  - The old full-sample load + training-side scatter path is removed.
  - The fetcher reconstructs local:
      - loss_mask
      - attention_mask
      - position_ids
  - USP batches require explicit attention_mask; non-USP batches keep the existing all-ones fallback.

  ### Collective Ordering Fix

  A USP local shard can have zero local loss tokens while its Ulysses peer has nonzero loss tokens. Previously, the zero-loss
  shard could skip the attention backward graph, while its peer still executed attention backward all-to-all collectives. That
  caused collective ordering divergence.

  This PR keeps the USP zero-loss path connected to hidden_states:

  local_sum_loss = local_sum_loss + hidden_states.sum() * 0.0

  This does not change the loss value, but preserves the same autograd collective sequence across USP ranks.
  
 ### USP attention correctness and microbenchmark

  Correctness is validated against `LlamaFlexAttention`:

  - `PYTHONPATH=. python -m unittest tests.test_usp_attention.TestUSPAttention`
  - Result: `Ran 2 tests in 50.9s`, `OK`

  | Mode | Config | Forward output max diff | Reduced loss diff | Projection grad max diff | Input grad max diff |
  | --- | --- | ---: | ---: | ---: | ---: |
  | Ulysses USP | `sp_ulysses_size=2, sp_ring_size=1` | `1.6e-1` | `6.7e-5` | `<= 7.0e-5` | `<= 1.9e-6` |
  | Ring USP | `sp_ulysses_size=1, sp_ring_size=2` | `1.8e-1` | `4.5e-5` | `<= 5.5e-5` | `<= 1.9e-6` |

  Attention-only microbenchmark setup:

  - `seq_len=8192`, `global_batch_size=2`
  - FlexAttention: 2-card DP, each rank processes `batch_size=1`
  - USP: 2-card SP, each global sequence is sharded across 2 ranks, with gradient accumulation `2`

  | Mode | Config | Median fwd+bwd time | Mean fwd+bwd time | Relative median time |
  | --- | --- | ---: | ---: | ---: |
  | FlexAttention DP | 2-card DP | `35.7-36.9 ms` | `36.0-37.2 ms` | `1.00x` |
  | Ulysses USP | `sp_ulysses_size=2, sp_ring_size=1` | `39.6 ms` | `40.9 ms` | `1.11x` |
  | Ring USP | `sp_ulysses_size=1, sp_ring_size=2` | `39.7 ms` | `39.5 ms` | `1.08x` |
    
  I also ran an end-to-end training comparison between the FlexAttention baseline and USP:

  - `llama31_8b_align_flex_match_usp`: FlexAttention baseline
  - `l31_usp_u2_bf16`: USP with `sp_ulysses_size=2`

  loss and accuracy comparison

  gradient norm comparison

  
  ## Limitations

  - Currently only SGLang supports producing USP-sharded Mooncake tensors.

### CODE REVIEW

**Comment by yubofredwang** [comment]
Thanks for the PR. Doing a verification locally

File: torchspec/models/draft/llama3_eagle.py [hunk 1]
```diff
@@ -1904,6 +1907,345 @@ def backward(ctx, grad_out):
...
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        cache_keys: Optional[torch.Tensor] = None,
+        cache_values: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        use_cache: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        del attention_mask, use_cache
```
[the remark this answers is not in the document]
Comment (uygnef) [reply torchspec/models/draft/llama3_eagle.py:2163]: This should be fine for the current path:  main attention is causal, so real tokens cannot attend to later padded tokens. The padded positions themselves are filtered by `loss_mask`, so they do not contribute to loss or gradients.

File: torchspec/models/eagle3.py [hunk 2]
```diff
@@ -139,19 +145,32 @@ def forward(
...
+                    f"USP local seq_length ({seq_length}) must be larger than ttt_length ({self.length})"
+                )
+            if position_ids is None:
+                device = hidden_states.device
+                position_ids = torch.arange(
+                    past_key_values_length,
+                    usp_chunk_size * self._usp_ulysses_world_size + past_key_values_length,
+                    dtype=torch.long,
+                    device=device,
+                ).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (uygnef) [reply torchspec/models/eagle3.py:175]: 👍 fixed

**Review by yubofredwang — APPROVED** [review]
